prototype/mobile_endpoint/dao.py

```diff
@@ -0,0 +1,212 @@
+from abc import ABCMeta, abstractmethod
+import functools
+import types
+
+import redis
+from mobile_endpoint.case.models import CommCareCase
+
+from mobile_endpoint.exceptions import IllegalCaseId
+from mobile_endpoint.form.models import XFormInstance
+from mobile_endpoint.models import db, Synclog, FormData, CaseData, CaseIndex
+from mobile_endpoint.utils import chunked, get_with_lock
+
+
+def to_generic(fn):
+    def _wrap(obj):
+        if hasattr(obj, 'to_generic'):
+            return obj.to_generic()
+        else:
+            return obj
+
+    @functools.wraps(fn)
+    def _inner(*args, **kwargs):
+        obj = fn(*args, **kwargs)
+        if isinstance(obj, (list, tuple)):
+            return [_wrap(ob) for ob in obj]
+        elif isinstance(obj, types.GeneratorType):
+            return (_wrap(ob) for ob in obj)
+        else:
+            return _wrap(obj)
+
+    return _inner
+
+class AbsctractDao(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def commit(self, xform, cases):
+        """
+        Commit the transaction
+        """
+        pass
+
+    @abstractmethod
+    def get_synclog(self, id):
+        pass
+
+    @abstractmethod
+    def get_form(self, id):
+        pass
+    
+    @abstractmethod
+    def get_case(self, id, lock=False):
+        pass
+
+    @abstractmethod
+    def case_exists(self, id):
+        pass
+
+    @abstractmethod
+    def iter_cases(self, case_ids, chunksize=100):
+        pass
+
+
+class SQLDao(AbsctractDao):
+    def commit(self, xform, cases):
```

Comment:
This is where the `XFormInstance` and `CommCareCase` objects get converted to SQL models and saved to the DB.